src/plots/cartesian/axes.js

```diff
@@ -777,20 +781,31 @@ axes.calcTicks = function calcTicks(ax, opts) {
             var A = tickVals[a].value;
             var B = tickVals[b].value;
 
-            var delta = definedDelta || Math.abs(B - A);
-            if(delta >= ONEDAY * 365) { // Years could have days less than ONEAVGYEAR period
-                v += ONEAVGYEAR / 2;
-            } else if(delta >= ONEAVGQUARTER) {
-                v += ONEAVGQUARTER / 2;
-            } else if(delta >= ONEMINMONTH) { // Months could have days less than ONEAVGMONTH period
-                var actualDelta = Math.abs(B - A);
+            var actualDelta = Math.abs(B - A);
+            var delta = definedDelta || actualDelta;
+            if(delta >= ONEMINYEAR) {
+                if(actualDelta >= ONEMINYEAR && actualDelta <= ONEMAXYEAR) {
+                    v += actualDelta / 2;
+                } else {
+                    v += ONEAVGYEAR / 2;
+                }
+            } else if(delta >= ONEMINQUARTER) {
+                if(actualDelta >= ONEMINQUARTER && actualDelta <= ONEMAXQUARTER) {
+                    v += actualDelta / 2;
+                } else {
+                    v += ONEAVGQUARTER / 2;
+                }
+            } else if(delta >= ONEMINMONTH) {
                 if(actualDelta >= ONEMINMONTH && actualDelta <= ONEMAXMONTH) {
                     v += actualDelta / 2;
                 } else {
                     v += ONEAVGMONTH / 2;
                 }
             } else if(delta >= ONEWEEK) {
                 v += ONEWEEK / 2;
+                if(actualDelta === ONEWEEK && ax._hasDayOfWeekBreaks) {
+                    v -= ONEDAY; // half of two days which is a good approximate for the number of week-end days
```

Comment:
I forget, did we already shift the ticks to look like they're right at the end of the rangebreak? If so (please confirm) this seems like it'll make sense for 5-day work weeks regardless of which two days were removed and whether weeks start on Sunday or Monday. But there are still a few cases I'm worried about:
- removing just one weekend day (eg Saturday is still included, just Sunday is removed)
- weekend AND nightly breaks - the shift wouldn't be correct, and could even result in a broken tick because it lands within another rangebreak
- holidays that shorten to a 4-day work week

The only thing I can think of is perhaps figure out the *actual* range of the week between A and B, go to the midpoint of THAT, and then check if the result is in a break and needs further shifting. Is that possible?

Reply:
> I forget, did we already shift the ticks to look like they're right at the end of the rangebreak? If so (please confirm)

Yes. The logic for setting `rangebreaks` labels is just above this part: 
https://github.com/plotly/plotly.js/blob/8928851519e1b74dda977cc574abaa9162820d2c/src/plots/cartesian/axes.js#L640-L682

Reply:
> there are still a few cases I'm worried about:
> 
> * removing just one weekend day (eg Saturday is still included, just Sunday is removed)
> * weekend AND nightly breaks - the shift wouldn't be correct, and could even result in a broken tick because it lands within another rangebreak
> * holidays that shorten to a 4-day work week
> 
> The only thing I can think of is perhaps figure out the _actual_ range of the week between A and B, go to the midpoint of THAT, and then check if the result is in a break and needs further shifting. Is that possible?

Good call. I am working on a commit to help fix these cases.

Reply:
Addressed in https://github.com/plotly/plotly.js/pull/5089/commits/f907f2a3f8009f2180d7bcf00bcd31e17c4351f8.

Reply:
> Weekend AND nightly breaks is still a problem: https://codepen.io/alexcjohnson/pen/yLOVjJy?editors=0010
> <img alt="Screen Shot 2020-08-21 at 12 04 30 PM" width="826" src="https://user-images.githubusercontent.com/2678795/90911194-9c06f780-e3a6-11ea-8cb3-b8146bd9dc75.png">

Good catch. This one addressed in https://github.com/plotly/plotly.js/pull/5089/commits/50a34519913d39339e15a70edbe566d5b73a1bdd.
Here is a [codepen](https://codepen.io/MojtabaSamimi/pen/poyrRQP?editors=0010).